storage/migrations/02_oasis_3_emerald.up.sql

```diff
@@ -204,6 +204,30 @@ CREATE TABLE oasis_3.emerald_withdraws
 CREATE INDEX ix_emerald_withdraws_sender ON oasis_3.emerald_withdraws(sender);
 CREATE INDEX ix_emerald_withdraws_receiver ON oasis_3.emerald_withdraws(receiver);
 
+-- Events emitted from the emerald paratime
+CREATE TABLE oasis_3.emerald_events
+(
+  round UINT63 NOT NULL REFERENCES oasis_3.emerald_rounds DEFERRABLE INITALLY DEFERRED,
+  tx_index UINT31,
+
+  tx_hash HEX64,
+  -- TODO: add link to openapi spec section with runtime event types.
+  type TEXT NOT NULL, 
+  -- The raw event, as returned by the oasis-sdk runtime client. This field is optional 
+  -- and will only exist for events from the `contracts` or `evm` modules, see
+  -- [contracts](https://pkg.go.dev/github.com/oasisprotocol/oasis-sdk/client-sdk/go@v0.2.0/modules/contracts#Event)
+  -- [evm](https://pkg.go.dev/github.com/oasisprotocol/oasis-sdk/client-sdk/go@v0.2.0/modules/evm#Event)
+  raw JSON, -- Raw event byte blob
+  -- The parsed event. Events from the `accounts`, `consensusaccounts`, and `core` modules are unchanged.
+  -- Events from the `evm` and `contracts` modules are parsed into known event types, e.g. ERC20.transfer.
+  -- If the raw event cannot be decoded, this field is empty.
+  body JSON, -- Parsed event
```

Comment:
Nope we don't _need_ `raw` but it sounds like what you're suggesting is similar? Essentially the `raw` would now be stored in `body` and the parsed `evm` event in `body` would be stored in `evm_log_params` instead.

My original thinking was that `raw` would store the raw Event struct returned by the oasis-sdk `runtimeClient`, and that `body` would store the processed version of that event after being processed by the indexer. The indexer would define its own custom structs for possible `contract` or `evm` events and store those in `body`. When queried for event info, the indexer would only look at the `body` col and return info about events it knows how to parse. This should make queries simpler.

I'd prefer flattening the various module `Event` types with the `Evm` events because most users will want to see the `evm` event types they're familiar with. However, it doesn't necessarily mean we have to store it this way in the database, and I agree that what you're suggesting better mirrors the 2-layered hierarchy of runtime events.

Reply:
I think we've agreed on the format pretty well on the call and in slack, so I'll keep this shorter.

For posterity, here's the excellent summary you just posted and that touches on several of the points above:

> We won’t be flattening the event types bc 1) many potential event types once we have contract verification, 2) it’s confusing to have raw evm.Log events when we cannot parse the event but also parsed evm.Transfer event on the same level, 3) it doesn’t reflect the 2-level event hierarchy well.

Also, about this:

> better mirrors the 2-layered hierarchy of runtime events

I do think we should aim for this. Makes it easier to understand the output for power users that are familiar with (or trying to wrap their head around) the internals of the oasis blockchain. (Probably the great majority of our consumers are power users; non-power users will use the GUI, and the GUI engineers are power users themselves and will IMO like the rawer presentation better too.) Also, makes it easier for us to describe the output.